tools/queries.cpp

```diff
@@ -43,86 +46,149 @@
-                uint64_t result = query_func(query, thresholds[idx]);
+                uint64_t result = query_func(query, thresholds[query_idx]);
                 if (safe && result < k) {
-                    num_reruns += 1;
+                    corrective_rerun_count += 1;
                     result = query_func(query, 0);
                 }
                 do_not_optimize_away(result);
             });
             if (run != 0) {  // first run is not timed
-                query_times.push_back(usecs.count());
+                times_per_query[query_idx][run - 1] = usecs.count();
             }
-            idx += 1;
+            query_idx += 1;
         }
     }
 
-    if (false) {
-        for (auto t: query_times) {
-            std::cout << (t / 1000) << std::endl;
-        }
-    } else {
-        std::sort(query_times.begin(), query_times.end());
-        double avg =
+    // Print JSON summary
+    nlohmann::json summary;
+    summary["encoding"] = index_type;
+    summary["algorithm"] = query_type;
+    summary["runs"] = runs;
+    summary["k"] = k;
+    summary["safe"] = safe;
+    summary["corrective_reruns"] = corrective_rerun_count;
+    summary["times"] = nlohmann::json::array();
+
+    auto add_aggregated_query_times = [&](AggregationType agg_type) {
+        auto query_times = aggregate_and_sort_times_per_query(agg_type, times_per_query);
+        auto agg_name = to_string(agg_type);
+
+        double mean =
             std::accumulate(query_times.begin(), query_times.end(), double()) / query_times.size();
         double q50 = query_times[query_times.size() / 2];
         double q90 = query_times[90 * query_times.size() / 100];
         double q95 = query_times[95 * query_times.size() / 100];
         double q99 = query_times[99 * query_times.size() / 100];
 
-        spdlog::info("---- {} {}", index_type, query_type);
-        spdlog::info("Mean: {}", avg);
-        spdlog::info("50% quantile: {}", q50);
-        spdlog::info("90% quantile: {}", q90);
-        spdlog::info("95% quantile: {}", q95);
-        spdlog::info("99% quantile: {}", q99);
-        spdlog::info("Num. reruns: {}", num_reruns);
+        summary["times"].push_back(
+            {{"query_aggregation", agg_name},
+             {"mean", mean},
+             {"q50", q50},
+             {"q90", q90},
+             {"q95", q95},
+             {"q99", q99}}
+        );
+    };
+
+    add_aggregated_query_times(AggregationType::None);
+    add_aggregated_query_times(AggregationType::Min);
+    add_aggregated_query_times(AggregationType::Mean);
+    add_aggregated_query_times(AggregationType::Median);
+    add_aggregated_query_times(AggregationType::Max);
+    std::cout << summary.dump(2) << "\n";
 
-        stats_line()("type", index_type)("query", query_type)("avg", avg)("q50", q50)("q90", q90)(
-            "q95", q95)("q99", q99);
+    // Save times per query (if required)
+    if (os != nullptr) {
```

Comment:
As mentioned above, I'd like to isolate this from extracting the data itself.

Once that's done, it should look something like this:

```
if (should_print) {
    auto os = open_file()
    // write results
}
```

We do away with a potentially-null pointer altogether.